pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/ReplicaGroupRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,573 @@
+    int targetNumReplicaGroup = tableConfig.getValidationConfig().getReplicationNumber();
+
+    OfflineTagConfig offlineTagConfig = new OfflineTagConfig(tableConfig, _helixManager);
+    List<String> serverInstances =
+        _helixAdmin.getInstancesInClusterWithTag(_helixClusterName, offlineTagConfig.getOfflineServerTag());
+
+    // Perform the basic validation
+    if (targetNumReplicaGroup <= 0 || targetNumInstancesPerPartition <= 0
+        || targetNumReplicaGroup * targetNumInstancesPerPartition > serverInstances.size()) {
+      throw new UnsupportedOperationException(
+          "Unsupported input config (numReplicaGroup: " + targetNumReplicaGroup + ", " + "numInstancesPerPartition: "
+              + targetNumInstancesPerPartition + ", numServers: " + serverInstances.size() + ")");
+    }
+
+    // If bootstrap mode is enabled, build the replica group partition assignment without checking rebalance mode.
+    if (rebalanceGroupBootstrap) {
+      return partitionAssignmentGenerator.buildReplicaGroupPartitionAssignment(tableConfig.getTableName(), tableConfig,
+          serverInstances);
+    }
+
+    // Fetch the previous replica group configurations
+    List<String> oldServerInstances = oldReplicaGroupPartitionAssignment.getAllInstances();
+    int oldNumReplicaGroup = oldReplicaGroupPartitionAssignment.getNumReplicaGroups();
+    int oldNumInstancesPerPartition = oldServerInstances.size() / oldNumReplicaGroup;
+
+    // Compute added and removed servers
+    List<String> addedServers = new ArrayList<>();
+    List<String> removedServers = new ArrayList<>();
+    for (String server : oldServerInstances) {
+      if (!serverInstances.contains(server)) {
+        removedServers.add(server);
+      }
+    }
+    for (String server : serverInstances) {
+      if (!oldServerInstances.contains(server)) {
+        addedServers.add(server);
+      }
+    }
+
+    // Create the new replica group partition assignment
+    ReplicaGroupPartitionAssignment newReplicaGroupPartitionAssignment =
+        new ReplicaGroupPartitionAssignment(oldReplicaGroupPartitionAssignment.getTableName());
+
+    // In case of replacement, create the removed to added server mapping
+    Map<String, String> oldToNewServerMapping = new HashMap<>();
+    if (addedServers.size() == removedServers.size()) {
+      for (int i = 0; i < addedServers.size(); i++) {
+        oldToNewServerMapping.put(removedServers.get(i), addedServers.get(i));
+      }
+    }
+
+    // Compute rebalance type
+    ReplicaGroupRebalanceType rebalanceType =
+        computeRebalanceType(oldNumInstancesPerPartition, oldNumReplicaGroup, targetNumReplicaGroup,
+            targetNumInstancesPerPartition, addedServers, removedServers);
+
+    // Based on the rebalance type, compute the new replica group partition assignment accordingly
+    for (int partitionId = 0; partitionId < oldReplicaGroupPartitionAssignment.getNumPartitions(); partitionId++) {
```

Comment:
I suppose increasing/decreasing number of partitions is not allowed, or currently unsupported? Good to add a comment here

Reply:
added comment